File: code-snippets/c++/oops/poly.cpp
#include<bits/stdc++.h>
using namespace std;

class Country {
public:
    virtual string capital() {
        return "CAPITAL";
    }
};

class India: public Country {
public:
    string capital() {
        return "Delhi";
    }
};

class USA: public Country {
public:
    string capital() {
        return "Washington DC";
    }
};

void print_capital(Country * c) {
    cout << c->capital() << endl;
}

int main() {
    Country * c;
    India obj1;

    c = &obj1;

    print_capital(c);

    USA * obj2;
    print_capital(obj2);
}

File: code-snippets/python/oops-and-basics/polymorphism/poly.py

class Country(object):
    """"""
    def capital(self): pass
    def language(self): pass
    def type(self): pass

class India(Country):
    def capital(self):
        print("New Delhi is the capital of India.")

    def language(self):
        print("Hindi is the most widely spoken language of India.")

    def type(self):
        print("India is a developing country.")

class USA(Country):
    def capital(self):
        print("Washington, D.C. is the capital of USA.")

    def language(self):
        print("English is the primary language of USA.")

    def type(self):
        print("USA is a developed country.")

def func(obj: Country) -> None:
    obj.capital()
    obj.language()
    obj.type()

obj_ind = India()
obj_usa = USA()

func(obj_ind)
func(obj_usa)
